winbuild/dir.cpp

```diff
@@ -48,22 +50,104 @@ class WinDirHandle : public watchman_dir_handle {
   }
 
   const watchman_dir_ent* readDir() override {
-    if (!info_) {
-      if (!GetFileInformationByHandleEx(
-              (HANDLE)h_.handle(), FileFullDirectoryInfo, buf_, sizeof(buf_))) {
-        if (GetLastError() == ERROR_NO_MORE_FILES) {
-          return nullptr;
+    if ((getenv("WATCHMAN_WIN7_COMPAT") && 
```

Comment:
Please put this detection logic in the constructor and set a `bool win7_` member variable.
It's also not super clear how this is detected at runtime without these env vars being set.  The env vars are used elsewhere to force this code path in the tests, but we also need a way to catch this and switch on `win7_` when these env vars are not set.

One approach for this is to try the win8 function, then if we get an error code that tells us it isn't implemented, set the win7 flag and try the other method.